### PR TITLE
Models registry

### PR DESCRIPTION
**Pull request by dakinggg**
Round of manual tests:

- [x] MPT pretrain before and after (`registry-condition-mpt-1-YoWX7j` vs `registry-baseline-mpt-1-gTfymO`)
- [x] llama finetune before and after (`registry-condition-llama-1-qxchzh` vs `registry-baseline-llama-1-eYIlHX`)
- [x] multi model eval before and after (`registry-condition-eval-1-dXlla3` vs `registry-baseline-eval-1-9Csev8`)

### CODE REVIEW

File: llmfoundry/utils/warnings.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def __init__(self, feature_name: str) -> None:
 
 
 # Decorator version of experimental warning
-def experimental(feature_name: str) -> Callable[[F], F]:
+def experimental_function(feature_name: str) -> Callable[[F], F]:
```
Comment (irenedea) [inline llmfoundry/utils/warnings.py:54]: Can we combine to still be a single function with a simple name `experimental`?
```
def experimental(feature_name: str) -> Callable[[F], F]:
    """Class decorator to mark a class as experimental."""
    def decorator(target: Union[Type, Callable]):
         if isinstance(target, type):
              return decorate_class(target)
        elif callable(target):
              return decorate_fn(target)
        raise Exception()
     return decorator
```

File: llmfoundry/utils/builders.py [hunk 2]
```diff
@@ -155,6 +159,35 @@ def build_icl_data_and_gauntlet(
...
+            registry=registry.models,
+            pre_validation_function=ComposerModel,
+            post_validation_function=None,
+            kwargs={
+                'om_model_config': cfg,
+                'tokenizer': tokenizer
+            },
+        )
+
+    if cfg.get('master_weights_dtype') in ('bf16', 'bfloat16'):
```
Comment (irenedea) [inline llmfoundry/utils/builders.py:183]: nit: pass in the `master_weights_dtype` as a direct arg to `build_composer_model`

File: llmfoundry/utils/builders.py [hunk 3]
```diff
@@ -155,6 +159,35 @@ def build_icl_data_and_gauntlet(
     return icl_evaluators, logger_keys, eval_gauntlet_cb
 
 
+def build_composer_model(
+    name: str,
+    cfg: DictConfig,
+    tokenizer: PreTrainedTokenizerBase,
+    init_context: Optional[ContextManager] = None,
+) -> ComposerModel:
```
Comment (irenedea) [inline llmfoundry/utils/builders.py:167]: doc string?

File: llmfoundry/models/inference_api_wrapper/interface.py [hunk 4]
```diff
@@ -1,22 +1,24 @@
...
 
 from llmfoundry.metrics import DEFAULT_CAUSAL_LM_EVAL_METRICS
-from llmfoundry.utils.builders import build_metric
 
 
 class InferenceAPIEvalWrapper(ComposerModel):
 
-    def __init__(self, model_cfg: Dict, tokenizer: AutoTokenizer):
+    def __init__(self, om_model_config: DictConfig, tokenizer: AutoTokenizer):
+        from llmfoundry.utils.builders import build_metric
```
Comment (irenedea) [inline llmfoundry/models/inference_api_wrapper/interface.py:20]: why?

**Review by irenedea — APPROVED** [review]
LGTM! some minor things

File: llmfoundry/utils/warnings.py (hunk 1 above)
Comment (dakinggg) [reply llmfoundry/utils/warnings.py:54]: I find this a little bit less clear with the typing so I am going to leave it as is

File: llmfoundry/models/inference_api_wrapper/interface.py (hunk 4 above)
Comment (dakinggg) [reply llmfoundry/models/inference_api_wrapper/interface.py:20]: circular imports

File: llmfoundry/utils/builders.py (hunk 2 above)
Comment (dakinggg) [reply llmfoundry/utils/builders.py:183]: sure, i guess i can pop them off and start down the config improvement road

File: llmfoundry/__init__.py [hunk 5]
```diff
@@ -31,7 +31,6 @@
     flash_attn_fn, scaled_multihead_dot_product_attention, triton_flash_attn_fn)
 from llmfoundry.models.layers.blocks import MPTBlock
 from llmfoundry.models.layers.ffn import FFN_CLASS_REGISTRY, MPTMLP, build_ffn
-from llmfoundry.models.model_registry import COMPOSER_MODEL_REGISTRY
```
Comment (dakinggg) [inline llmfoundry/__init__.py:34]: something wrong, blocking merge while resolving

Comment (dakinggg) [reply llmfoundry/__init__.py:34]: resolved

**Pull request merged**